### DISCUSSION TITLE
Add audio stream from usb mic to Raspberrycam in docker contener

### DISCUSSION THREAD

**Discussion by Topper69** [bluenviron/mediamtx#1425] [Questions] [answered]
Hi all,

Many thanks for this project, it's awsome!

Q1:
Do you have a solution to add the audio stream from usb mic to the raspberry pi camera feed?

Plateform:
Raspberry pi 4 @ Debian bulleseye
rtsp-simple-server @ docker aarch64 image

Q2:
Do you have a tips to customize Gstream pipeline to something like that :
gst-launch-1.0 libcamerasrc ! video/x-raw, width=640, height=480, framerate=30/1 ! videoconvert ! videoscale ! clockoverlay time-format="%D %H:%M:%S" ! autovideosink

NOTA : gstream is broken on the lastest version of bulleseye 64b (1.18.4-2 has we speek). It segfault when we use with libcamerasrc.

Q3:
Any plan to add --tuning-file option for ex to use with the NoIR camera (/usr/share/libcamera/ipa/raspberrypi/imx219_noir.json)

Many thanks for the help.

Best regards

**Comment by Topper69** [comment, marked as answer]
Answers for public knowledge:

Q1:
I feed video + audio with gstreamer pipeline : 
on the config file : 
path:
  mystream:
    runOnDemand: gst-launch-1.0 rtspclientsink name=s location=rtsp://localhost:8554/mystream libcamerasrc ! video/x-raw, width=640, height=480, framerate=30/1 ! videoconvert ! videoscale ! clockoverlay time-format="%d/%m/%y %H:%M:%S" ! x264enc speed-preset=ultrafast bitrate=600 key-int-max=40 ! queue ! s.sink_0 alsasrc ! queue ! voaacenc ! s.sink_1
    runOnDemandRestart: yes

Q2 : see Q1 answer

To fix libcamera SIGFAULT issue, just update raspberry firmware with sudo rpi-update && sudo reboot

Q3 : with libcamera it'll be possible. With Gstreamer pipeline not sure.

**Reply by jarrah31** [reply]
Hi - I'm new to rtsp-simple-server and streamer, and I'm very keen to understand how you combine a Raspberry Pi Camera with an audio feed please.
Out of the various parameters you list above, which ones do the audio feed and how do you specify which USB mic to use? Is it done via a particular device file somehow?

**Reply by Topper69** [reply]
Hi,
The end part of the Gstreamer pipeline is the audio stream "alsasrc ! queue ! audioconvert ! opusenc ! s.sink_1"
In my case, i use alsasrc to record from my usb mic. You can also use pulseaudio, it depend on your wish and distrib you have.
I have also only one mic connected, so i don't have to specify the interface on my pipeline, it select by default the good one.

You can select your mic : ex "alsasrc device=hw:2 ! queue ! audioconvert  ..".
To list your cards, type "aplay -l"

Have fun.

**Reply by jarrah31** [reply]
Thank you @Topper69!

**Comment by aler9** [comment]
The GStreamer pipeline by @Topper69 was included in the README in an improved version, that doesn't use x264enc and doesn't consume CPU:

https://github.com/bluenviron/mediamtx#from-a-raspberry-pi-camera

[a comment by an automated account is not shown]
